Add overwriteRootCapability argument to capacity plugin

mrendi29 opened the pull request:
#### What type of PR is this?
Feature

#### What this PR does / why we need it:
Problem: 
Currently, the root queue's resource limits are always set to the actual cluster resources, regardless of any user-configured capability values. This prevents administrators from setting virtual resource limits that differ from the physical cluster capacity. This helps in cases where a user has cluster-autoscaled enabled and the "virtual" resources can be higher than the "real" resources. 

#### Which issue(s) this PR fixes:
Fixes #3910 

#### Special notes for your reviewer: I have not yet tested this in our production clusters, I plan to test it in the following week. 

#### Does this PR introduce a user-facing change?
None
```release-note
Add overwriteRootCapability argument to capacity plugin to allow using user-configured root queue capability instead of actual cluster resources
```

mrendi29 commented:
cc: @JesseStutler @Monokaix

hajnalmt commented:
This is not true. You can set it to any value. We only set it to the cluster capability when it's empty.
https://github.com/volcano-sh/volcano/blob/master/pkg/scheduler/plugins/capacity/capacity.go#L590
Aren't you using an old version or something?

I am far from happy with a current queue design, and all these hacks around, but this is still not the right path.
I don't have time to implement it currently, but we should use overcommitFactors instead to align the configurations across the plugins...
https://github.com/volcano-sh/volcano/issues/4530

hajnalmt commented:
Oh sry, I understand it now. We want to overwrite the realCapability with the root Queue's capability, maybe renaming it to overwriteRootQueueRealCapability would better reflect the intention and logic behind the usage of the argument.

mrendi29 commented:
@hajnalmt thanks for the review. I'll update the variable to better reflect the intention

mrendi29 committed:
change name to overwriteRootQueueRealCapability and add more logging

Signed-off-by: Endi Caushi <mrendi29@gmail.com>

mrendi29 committed:
feat: capacity plugin overwriteRootQueueRealCapability

Introducing new optional argument in the capacity plugin
overwriteRootQueueRealCapability.

Currently, the root queue's resource limits are always set to the actual cluster
resources, regardless of any user-configured capability values.
This prevents administrators from setting virtual resource limits that differ from
the physical cluster capacity.

The new option helps in cases where a user has cluster-autoscaler enabled where
the "virtual" resources can be higher than the "real" resources.

Signed-off-by: Endi Caushi <mrendi29@gmail.com>

mrendi29 committed:
feat: const keys as arguments in capacity plugin

Based on the practices in other plugins, the overwriteRootQueueRealCapability
new argument in the capacity plugin comes from a const key.
This makes it easier to follow the configurable arguments of the plugin.

Signed-off-by: Endi Caushi <mrendi29@gmail.com>

mrendi29 commented:
@JesseStutler @hajnalmt  can you take a second look at this PR? I believe I have addressed all comments.  Let me know what you think. 

FYI, the test case is AI generated after a couple of prompts :)

JesseStutler commented on pkg/scheduler/plugins/capacity/capacity_test.go:
```diff
@@ -654,6 +654,205 @@ func Test_capacityPlugin_OnSessionOpenWithHierarchy(t *testing.T) {
 	}
 }
 
+func TestOverwriteRootQueueRealCapability(t *testing.T) {
```
I think we can just add a simple case in `Test_capacityPlugin_OnSessionOpenWithHierarchy` is enough, there's no need to add a new `Test_capacityPlugin_OnSessionOpenWithHierarchy` test func and add a lot of new cases :)

JesseStutler commented on pkg/scheduler/plugins/capacity/capacity.go:
```diff
@@ -593,7 +601,20 @@ func (cp *capacityPlugin) buildHierarchicalQueueAttrs(ssn *framework.Session) bo
+			rootQueueAttr.capability, cp.totalResource)
+	} else {
+		// Default behavior: use actual cluster resources
+		rootQueueAttr.realCapability = cp.totalResource.Clone()
```
@vzhou-p Looks like difference users have difference requirements, setting it as a infinite value is a good choice, but currrently child queues' capability will inherit from the parent queue if not set:
<img width="1207" height="214" alt="image" src="https://github.com/user-attachments/assets/c5722111-8efd-495b-99d7-e6c044378483" />
So which mean child queue also can submit infinite jobs, so I'm not sure that whether it's what users want, because difference users may have difference requirements

mrendi29 replied:
Setting the root queue as infinity sounds like a really interesting option. That means that I as a platform admin would not need to interact with the root queue directly and can just create another queue on top of root. 
(I recall that there was another PR somewhere that wanted to interact with the root queue capacity through the helm chart, this could eliminate that).

In the same time, having child queues submit infinite jobs seems really risky. 

I wonder though, if it makes sense for this functionality to be part of overwriteRootQueueRealCapability  in [L610](https://github.com/mrendi29/volcano/blob/f1f617d28db1d2fbac34949f8b3ed79c2e12e4ed/pkg/scheduler/plugins/capacity/capacity.go#L610) instead of default Volcano behavior. 

Depending on what the PMCs think I can then make the necessary adjustments.

mrendi29 commented:
If i submit the job to the root queue directly: 
Scheduler logs:
```
│ I1106 23:53:28.184662       1 util.go:101] schedulerPodName  is responsible to PodGroup spark/spark-spark-sleep-8-pg                                                       │
│ I1106 23:53:28.184747       1 event_handlers.go:835] Add PodGroup(spark-spark-sleep-8-pg) into cache, spec(v1beta1.PodGroupSpec{MinMember:1, MinTaskMember:map[string]int3 │
│ I1106 23:53:31.994400       1 util.go:63] schedulerPodName  is responsible to Pod spark/spark-sleep-8-driver                                                               │
│ I1106 23:53:31.994451       1 event_handlers.go:398] Added pod <spark/spark-sleep-8-driver> into cache.   
```

```
┌────────────────────────────────────────────────────────────────── Describe(spark/spark-spark-sleep-8-pg) ──────────────────────────────────────────────────────────────────┐
│ Name:         spark-spark-sleep-8-pg                                                                                                                                       │
│ Namespace:    spark                                                                                                                                                        │
│ Labels:       <none>                                                                                                                                                       │
│ Annotations:  <none>                                                                                                                                                       │
│ API Version:  scheduling.volcano.sh/v1beta1                                                                                                                                │
│ Kind:         PodGroup                                                                                                                                                     │
│ Metadata:                                                                                                                                                                  │
│   Creation Timestamp:  2025-11-06T23:53:28Z                                                                                                                                │
│   Generation:          1                                                                                                                                                   │
│   Owner References:                                                                                                                                                        │
│     API Version:           sparkoperator.k8s.io/v1beta2                                                                                                                    │
│     Block Owner Deletion:  true                                                                                                                                            │
│     Controller:            true                                                                                                                                            │
│     Kind:                  SparkApplication                                                                                                                                │
│     Name:                  spark-sleep-8                                                                                                                                   │
│     UID:                   a8ff063f-d444-4890-a75b-4be4b3bb5ad2                                                                                                            │
│   Resource Version:        221651212                                                                                                                                       │
│   UID:                     96c19dc5-4382-47e2-8a54-74f4564cd1ec                                                                                                            │
│ Spec:                                                                                                                                                                      │
│   Min Member:  1                                                                                                                                                           │
│   Min Resources:                                                                                                                                                           │
│     Cpu:     122                                                                                                                                                           │
│     Memory:  1208G                                                                                                                                                         │
│   Queue:     root                                                                                                                                                          │
│ Status:                                                                                                                                                                    │
│   Phase:  Pending                                                                                                                                                          │
│ Events:   <none>   
```

The driver is pending as well. 

Scheduler conf: 
```
actions: "enqueue, allocate, backfill"
tiers:
- plugins:
  - name: priority
  - name: gang
    enablePreemptable: false
  - name: conformance
- plugins:
  - name: drf
    enablePreemptable: false
  - name: predicates
  - name: capacity
    enableHierarchy: true 
    arguments:
      overwriteRootQueueRealCapability: true
  - name: nodeorder
  ```

JesseStutler commented:
> I tried testing this in my k8s cluster today. I created a new image only for the scheduler and used that in my cluster. I did not change controller and admission image. I created a root queue with 500 cores and 5000GB RAM (note my cluster has only <100 cores and 500GB RAM but has autoscaling enabled) I also created a child queue of root with 250 cores and 4500GB RAM.
> 
> I tried submitting a spark job but got the following error from the spark operator logs:
> 
> ```
> │ I1106 23:33:48.303393      10 event.go:285] Event(v1.ObjectReference{Kind:"SparkApplication", Namespace:"spark", Name:"spark-sleep-8", UID:"c3526aed-2566-42e1-a0a6-33a445cca51f", APIVersion:"sparkoperator.k8s.io/v1beta2", ResourceVersion:"221645563", FieldPath:""}): type: 'Normal' reason: 'SparkApplicationAdded' SparkApplication spark-sleep-8 was added, enqueuing it for submission                                                                                                                                │
> │ E1106 23:33:48.319225      10 controller.go:676] failed to process batch scheduler BeforeSubmitSparkApplication with error failed to sync PodGroup with error: admission webhook "validatepodgroup.volcano.sh" denied the request: can only submit PodGroup to queue with state `Open`, queue `myqueue` status is ``. Abandon schedule pods via volcano     
> ```
> 
> Seems that the queue i created is not reporting status:
> 
> ```
> ┌─────────────────────────────────────────────────────────────────────────── Describe(-/myqueue) ────────────────────────────────────────────────────────────────────────────┐
> │ Name:         myqueue                                                                                                                                                      │
> │ Namespace:                                                                                                                                                                 │
> │ Labels:       <none>                                                                                                                                                       │
> │ Annotations:  <none>                                                                                                                                                       │
> │ API Version:  scheduling.volcano.sh/v1beta1                                                                                                                                │
> │ Kind:         Queue                                                                                                                                                        │
> │ Metadata:                                                                                                                                                                  │
> │   Creation Timestamp:  2025-11-06T23:30:58Z                                                                                                                                │
> │   Generation:          1                                                                                                                                                   │
> │   Resource Version:    221644752                                                                                                                                           │
> │   UID:                 b9ad6d06-4ac8-4ca0-8e73-25964af8f0d6                                                                                                                │
> │ Spec:                                                                                                                                                                      │
> │   Guarantee:                                                                                                                                                               │
> │     Resource:                                                                                                                                                              │
> │       Cpu:      240                                                                                                                                                        │
> │       Memory:   4500G                                                                                                                                                      │
> │   Parent:       root                                                                                                                                                       │
> │   Reclaimable:  true                                                                                                                                                       │
> │   Weight:       1                                                                                                                                                          │
> │ Events:         <none>       
> ```
> 
> Whereas root:
> 
> ```
> │ Name:         root                                                                                                                                                         │
> │ Namespace:                                                                                                                                                                 │
> │ Labels:       <none>                                                                                                                                                       │
> │ Annotations:  <none>                                                                                                                                                       │
> │ API Version:  scheduling.volcano.sh/v1beta1                                                                                                                                │
> │ Kind:         Queue                                                                                                                                                        │
> │ Metadata:                                                                                                                                                                  │
> │   Creation Timestamp:  2025-08-20T22:08:49Z                                                                                                                                │
> │   Generation:          3                                                                                                                                                   │
> │   Resource Version:    188581782                                                                                                                                           │
> │   UID:                 de2d7990-f6dd-42c2-8e9d-524d6f07845b                                                                                                                │
> │ Spec:                                                                                                                                                                      │
> │   Capability:                                                                                                                                                              │
> │     Cpu:        500                                                                                                                                                        │
> │     Memory:     5000G                                                                                                                                                      │
> │   Reclaimable:  false                                                                                                                                                      │
> │   Weight:       1                                                                                                                                                          │
> │ Status:                                                                                                                                                                    │
> │   Allocated:                                                                                                                                                               │
> │     Cpu:     0                                                                                                                                                             │
> │     Memory:  0                                                                                                                                                             │
> │   Reservation:                                                                                                                                                             │
> │   State:  Open                                                                                                                                                             │
> │ Events:   <none>                                                                                                                                                           │
> │                    
> ```
> 
> Will do some more digging

It's weird, maybe we could check controller's log, I think there may be something wrong with controller to sync the state of queue from `""` to `Open`:
https://github.com/volcano-sh/volcano/blob/83f550edb4692ac0b3f499bc38b78705fbfb3e55/pkg/controllers/queue/state/open.go#L43-L48

hajnalmt committed:
feat: capacity plugin realCapability tests

Minor changes into the comments of the plugin.
Added a positive/negative unit test to test the argument.
Adding some informations about the argument to the capacity
plugin's hierarchy documentation.

Signed-off-by: Hajnal Mate <hajnalmt@gmail.com>

hajnalmt reviewed (approved):
/lgtm

Hello @mrendi29 ,
Sorry for some reason I thought this is already merged in.
I rebased your branch, modified the logging a little bit (thank you for the access). Also I removed the AI tests with a simple negative-positive test case that follows the structure of the existing test suite. Also added the argument with some context to the appropriate design doc.

I am not sure about the failure you got, but it shouldn't be due to these changes, and this argument can be useful in other use cases.

Please take a look @JesseStutler

mrendi29 commented:
@JesseStutler can you take a look at the current state of the PR? We can resolve conflict afterwards if the approach looks good to you

JesseStutler commented:
> @JesseStutler can you take a look at the current state of the PR? We can resolve conflict afterwards if the approach looks good to you
> 
> UPDATE: i also saw #4973 being worked on, that should also in theory address this

Yes, #4973 will set the capability of root queue to infinite by default, but we can also overwrite the capability, WDYT? Is this way better? @mrendi29 @hajnalmt

mrendi29 commented:
> I think this can be closed as https://github.com/volcano-sh/volcano/pull/4973 got merged in, as the root cause of the issue is solved with it. What do you think @mrendi29 ? 

I'll re-deploy volcano tomorrow in our QA cluster and double check that it solves the root cause. I believe it should, but i will let you know

mrendi29 commented:
Testing this next week, was busy with iceberg conference and few other projects
